community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundQueryContext.scala

```diff
@@ -586,34 +585,39 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
   }
 
   type KernelProcedureCall = (KernelQualifiedName, Array[AnyRef]) => RawIterator[Array[AnyRef], ProcedureException]
+  type KernelFunctionCall = (KernelQualifiedName, Array[AnyRef]) => AnyRef
+
+  private def shouldElevate(allowed: Array[String]): Boolean = {
+    // We have to be careful with elevation, since we cannot elevate permissions in a nested procedure call
+    // above the original allowed procedure mode. We enforce this by checking if mode is already an overridden mode.
+    val mode = transactionalContext.accessMode
+    allowed.nonEmpty && !mode.isOverridden && mode.getOriginalAccessMode.allowsProcedureWith(allowed)
```

Comment:
Yes, we only elevate access on the outermost procedure call. Once you got your elevated access rights, they can only be restricted by nested procedure calls. In the case of a WRITE procedure calling a READ procedure, it will get restricted to READ in the nested call. To be able to call the outer WRITE procedure you either need WRITE permissions or to be allowed WRITE permissions by annotation, which then implies READ permissions. Since the procedure modes have a strict order of which mode implies another, this should be sufficient.